message AuthInfo {
  message Simple {
    optional string username = 1;
    optional string password = 2;
  }

  optional Simple simple = 1;
}

message Request {
  optional string method_name = 1;
  optional bytes message = 2;
  optional uint64 timestamp = 3;
  optional uint64 deadline = 4;
}

message RpcError {
  message ExceptionInfo {
    optional bytes traceback = 1;
  }

  optional ExceptionInfo exception_info = 1;
}

message Response {
  optional bytes message = 1;
  optional uint32 streaming = 2;
  optional uint64 timing = 3;
}

message ProtocolMessage {
  optional uint32 sequence_number = 1;
  optional Request request = 2;
  optional Response response = 3;
  optional RpcError error = 4;
  optional AuthInfo auth_info = 5;
}

message LoggedProtocolMessage {
  optional uint64 timestamp = 1;
  optional ProtocolMessage sent = 2;
  optional ProtocolMessage received = 3;
}
